### PR TITLE
Add environment variables for I/O thread pool and slice sizes

### PR DESCRIPTION
**Pull request by vuule**
Modified `getenv_or` to allow the default value of any type.
Adds environment variables:

- `LIBCUDF_CUFILE_THREAD_COUNT`: integer, number of threads in the pool used for I/O (default 16);
- `LIBCUDF_CUFILE_SLICE_SIZE`: integer, maximum size of cuFile reads/writes in bytes (default 4MB).

### CODE REVIEW

**Review by devavret — APPROVED** [review]

File: cpp/src/io/utilities/config_utils.hpp [hunk 1]
```diff
@@ -23,7 +24,17 @@ namespace cudf::io::detail {
  * @brief Returns the value of the environment variable, or a default value if the variable is not
  * present.
  */
-std::string getenv_or(std::string const& env_var_name, std::string_view default_val);
```
Comment (ttnghia) [inline cpp/src/io/utilities/config_utils.hpp:26]: I feel something dangerous here. When you call `getenv_or("abc")`, you construct a temp string then in turn construct a `string_view` from it. So you are viewing a temp object which may not be guaranteed to exist in this function.

Comment (devavret) [reply cpp/src/io/utilities/config_utils.hpp:26]: I thought that was okay, [per cppref](https://en.cppreference.com/w/cpp/language/lifetime)
>All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception. 

Please correct me if wrong.

Comment (ttnghia) [reply cpp/src/io/utilities/config_utils.hpp:26]: Sorry I'm not sure what you mean?
To clarify what I said, the `string_view` parameter does not own the temp string. As such, the temp string may be destroyed immediately before entering the function.

Comment (ttnghia) [reply cpp/src/io/utilities/config_utils.hpp:26]: Okay, thank you. So it is safe to use in this case. I also found a long explanation for it here: https://www.foonathan.net/2017/03/string_view-temporary/

Comment (ttnghia) [reply cpp/src/io/utilities/config_utils.hpp:26]: I search over the internet and found a huge collection of the same topic: Is it safe to use `string_view` on temp string?

Comment (vuule) [reply cpp/src/io/utilities/config_utils.hpp:26]: AFAICT using `string_view` has the same pitfalls as const&. It should be fine as a parameter. I can change it back, don't have a strong preference.

Comment (ttnghia) [reply cpp/src/io/utilities/config_utils.hpp:26]: It's fine for me now.

**Review by ttnghia — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**